File: topics-ros2/scripts/turtle_controller/turtle_controller/turtle_controller.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from turtlesim.msg import Pose


class TurtleControllerNode(Node):
    """
    Nó controlador para a tartaruga no ambiente turtlesim.
    Publica no tópico de velocidade e recebe informações do tópico de posição (pose).
    """

    def __init__(self):
        """
        Inicializa o nó ROS 2 com o nome 'turtle_controller'.
        Cria o publicador para '/turtle1/cmd_vel' e o assinante (subscriber) para '/turtle1/pose'.
        """
        super().__init__("turtle_controller")
        self.cmd_vel_pub_ = self.create_publisher(Twist, "/turtle1/cmd_vel", 10)
        self.pose_sub_ = self.create_subscription(Pose, "/turtle1/pose", self.callback_pose, 10)

    def callback_pose(self, pose: Pose):
        """
        Função de callback chamada sempre que uma nova pose da tartaruga é recebida.
        Define e publica as velocidades linear e angular da tartaruga dependendo da sua coordenada 'x'.
        """
        cmd = Twist()
        if pose.x < 5.5:
            # Se a posição em x for menor que 5.5
            cmd.linear.x = 1.0
            cmd.angular.z = 1.0
        else:
            # Se a posição em x for maior ou igual a 5.5
            cmd.linear.x = 2.0
            cmd.angular.z = 2.0
        self.cmd_vel_pub_.publish(cmd)


def main(args=None):
    """
    Ponto de entrada principal do programa.
    Inicializa a comunicação ROS (rclpy), cria a instância do nó
    e o mantém em execução (spin) até que seja interrompido.
    """
    rclpy.init(args=args)
    node = TurtleControllerNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: topics-ros2/scripts/turtle_controller_cpp/src/turtle_controller.cpp
#include "geometry_msgs/msg/twist.hpp"
#include "rclcpp/rclcpp.hpp"
#include "turtlesim/msg/pose.hpp"

// Usado para facilitar a vinculação do parâmetro no callback
using namespace std::placeholders;

// Classe que define o nó ROS 2 para controlar a tartaruga
class TurtleControllerNode : public rclcpp::Node {
public:
  // Construtor do nó, define o nome do nó como "turtle_controller"
  // Nota: usando rclcpp::Node para evitar o erro de namespace ocorrido
  // anteriormente
  TurtleControllerNode() : rclcpp::Node("turtle_controller") {
    // Cria um publicador (publisher) no tópico "/turtle1/cmd_vel" para enviar
    // comandos de velocidade
    cmd_vel_pub_ = this->create_publisher<geometry_msgs::msg::Twist>(
        "/turtle1/cmd_vel", 10);

    // Cria um assinante (subscriber) no tópico "/turtle1/pose" para receber a
    // posição atual da tartaruga
    pose_sub_ = this->create_subscription<turtlesim::msg::Pose>(
        "/turtle1/pose", 10,
        std::bind(&TurtleControllerNode::poseCallback, this, _1));
  }

  // Função de callback acionada sempre que uma nova mensagem de posição é
  // recebida
  void poseCallback(const turtlesim::msg::Pose::SharedPtr pose) {
    auto cmd = geometry_msgs::msg::Twist();

    // Verifica se a tartaruga está na metade esquerda (x < 5.5) ou direita (x
    // >= 5.5) da tela
    if (pose->x < 5.5) {
      // Se estiver na parte esquerda, move-se mais devagar
      cmd.linear.x = 1.0;
      cmd.angular.z = 1.0;
    } else {
      // Se estiver na parte direita, move-se mais rápido
      cmd.linear.x = 2.0;
      cmd.angular.z = 2.0;
    }

    // Publica a mensagem de comando de velocidade
    cmd_vel_pub_->publish(cmd);
  }

private:
  // Ponteiros compartilhados para o publicador de velocidade e o assinante de
  // posição
  rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr cmd_vel_pub_;
  rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr pose_sub_;
};

int main(int argc, char **argv) {
  // Inicializa a infraestrutura de comunicação do ROS 2
  rclcpp::init(argc, argv);

  // Cria uma alocação de memória inteligente (smart pointer) para a instância
  // do nó
  auto node = std::make_shared<TurtleControllerNode>();

  // Mantém o nó em execução contínua, processando os callbacks de eventos
  rclcpp::spin(node);

  // Finaliza adequadamente a execução do ROS 2 antes de encerrar o programa
  rclcpp::shutdown();
  return 0;
}